src/transforms/hexbin.js

```diff
@@ -10,52 +11,99 @@ import {hasOutput, maybeOutputs} from "./group.js";
 export const ox = 0.5, oy = 0;
 
 export function hexbin(outputs = {fill: "count"}, options = {}) {
-  const {radius, ...rest} = outputs;
-  return hexbinn(rest, {radius, ...options});
+  const {binWidth, ...rest} = outputs;
+  return hexbinn(rest, {binWidth, ...options});
 }
 
-// TODO group by (implicit) z
 // TODO filter e.g. to show empty hexbins?
-// TODO data output with sort and reverse?
 // TODO disallow x, x1, x2, y, y1, y2 reducers?
-function hexbinn(outputs, {radius = 10, ...options}) {
-  radius = +radius;
-  outputs = maybeOutputs(outputs, options);
+function hexbinn(outputs, {binWidth = 20, fill, stroke, z, ...options}) {
+  binWidth = +binWidth;
+  const [GZ, setGZ] = maybeChannel(z);
+  const [vfill] = maybeColorChannel(fill);
+  const [vstroke] = maybeColorChannel(stroke);
+  const [GF = fill, setGF] = maybeChannel(vfill);
+  const [GS = stroke, setGS] = maybeChannel(vstroke);
+  outputs = maybeOutputs({
+    ...setGF && {fill: "first"},
+    ...setGS && {stroke: "first"},
+    ...outputs
+  }, {fill, stroke, ...options});
   return {
     symbol: "hexagon",
-    ...!hasOutput(outputs, "r") && {r: radius},
-    ...hasOutput(outputs, "fill") && {stroke: "none"},
+    ...!hasOutput(outputs, "r") && {r: binWidth / 2},
+    ...!setGF && {fill},
+    ...((hasOutput(outputs, "fill") || setGF) && stroke === undefined) ? {stroke: "none"} : {stroke},
     ...options,
     initialize(data, facets, {x: X, y: Y}, {x, y}) {
+      if (setGF) setGF(valueof(data, vfill));
+      if (setGS) setGS(valueof(data, vstroke));
+      if (setGZ) setGZ(valueof(data, z));
+      for (const o of outputs) o.initialize(data);
       if (X === undefined) throw new Error("missing channel: x");
       if (Y === undefined) throw new Error("missing channel: y");
-      X = X.value;
-      Y = Y.value;
-      const binsof = Hexbin().x(i => x(X[i]) - ox).y(i => y(Y[i]) - oy).radius(radius * sqrt4_3);
+      X = X.value.map(x);
+      Y = Y.value.map(y);
```

Comment:
It isn’t safe to use X.value.map here because _array_.map returns the same type of _array_ which may be a typed array such as Uint8Array. Or to put it another way, X.value is in data space, whereas the result of applying the _x_ scale is in screen space and can be any number (such as Float64Array).